### PR TITLE
Fix location for SQLi when using mysql Pool

### PR DESCRIPTION
**Pull request by CarlesDD**
### What does this PR do?
Fixes the location for detected SQLi when code base uses mysql Pool to query the DB.

### Motivation
Provide the correct location for detected vulnerabilities.

### Plugin Checklist
- [x] Unit tests.

### CODE REVIEW

File: packages/datadog-instrumentations/src/mysql.js [hunk 1]
```diff
@@ -66,9 +66,49 @@ addHook({ name: 'mysql', file: 'lib/Connection.js', versions: ['>=2'] }, Connect
 })
 
 addHook({ name: 'mysql', file: 'lib/Pool.js', versions: ['>=2'] }, Pool => {
+  const startPoolQueryCh = channel('datadog:mysql:pool:query:start')
+  const finishPoolQueryCh = channel('datadog:mysql:pool:query:finish')
```
Comment (uurien) [inline packages/datadog-instrumentations/src/mysql.js:70]:
```suggestion
  const finishPoolQueryCh = channel('datadog:mysql:pool:query:finish')

```

File: packages/datadog-instrumentations/src/mysql.js [hunk 2]
```diff
@@ -66,9 +66,49 @@ addHook({ name: 'mysql', file: 'lib/Connection.js', versions: ['>=2'] }, Connect
...
+        retval.then(() => {
+          finish()
+        }).catch(() => {
+          finish()
+        })
+      }
+
+      return retval
+    })
+  })
```
Comment (uurien) [inline packages/datadog-instrumentations/src/mysql.js:112]:
```suggestion
  })

```

File: packages/dd-trace/src/appsec/iast/analyzers/sql-injection-analyzer.js [hunk 3]
```diff
@@ -20,37 +20,33 @@ class SqlInjectionAnalyzer extends InjectionAnalyzer {
...
-      const store = storage.getStore()
-      if (store && store.pgPoolParentStore) {
-        storage.enterWith(store.pgPoolParentStore)
-      }
-    })
+    this.addSub(
+      'datadog:sequelize:query:start',
+      ({ sql, dialect }) => this.getStoreAndAnalyze(sql, dialect.toUpperCase())
+    )
+    this.addSub('datadog:sequelize:query:finish', this.returnToParentStore)
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/analyzers/sql-injection-analyzer.js:27]: With the current approach, youare replacing the `this` property of `returnToParentStore` to unknown one.
```suggestion
    this.addSub('datadog:sequelize:query:finish', () => this.returnToParentStore())
```

Comment (uurien) [reply packages/dd-trace/src/appsec/iast/analyzers/sql-injection-analyzer.js:27]: same in L33

File: packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js [hunk 4]
```diff
@@ -8,52 +10,121 @@ const vulnerabilityReporter = require('../../../../src/appsec/iast/vulnerability
...
-              resolve()
-            }
+
+        testThatRequestHasVulnerability(() => {
+          return new Promise((resolve, reject) => {
+            const store = storage.getStore()
+            const iastCtx = iastContextFunctions.getIastContext(store)
+            let sql = 'SELECT 1'
+            sql = newTaintedString(iastCtx, sql, 'param', 'Request')
+            queryMethods.executeQueryWithCallback(sql, connection, function (err) {
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:53]:
```suggestion

            queryMethods.executeQueryWithCallback(sql, connection, function (err) {
```

File: packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js [hunk 5]
```diff
@@ -8,52 +10,121 @@ const vulnerabilityReporter = require('../../../../src/appsec/iast/vulnerability
...
+          pool.end(done)
+        })
+
+        testThatRequestHasVulnerability(() => {
+          return new Promise((resolve, reject) => {
+            const store = storage.getStore()
+            const iastCtx = iastContextFunctions.getIastContext(store)
+            let sql = 'SELECT 1'
+            sql = newTaintedString(iastCtx, sql, 'param', 'Request')
+            queryMethods.executeQueryWithCallback(sql, pool, function (err) {
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:102]:
```suggestion

            queryMethods.executeQueryWithCallback(sql, pool, function (err) {
```

File: packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 'use strict'
-
+const fs = require('fs')
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:2]: Sorry, I didn't see this in the first review
```suggestion

const fs = require('fs')
```

File: packages/datadog-instrumentations/src/mysql.js [hunk 7]
```diff
@@ -66,9 +66,51 @@ addHook({ name: 'mysql', file: 'lib/Connection.js', versions: ['>=2'] }, Connect
...
   })
+
+  shimmer.wrap(Pool.prototype, 'query', query => function () {
+    if (!startPoolQueryCh.hasSubscribers) {
+      return query.apply(this, arguments)
+    }
+
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+    const sql = arguments[0].sql ? arguments[0].sql : arguments[0]
```
Comment (simon-id) [inline packages/datadog-instrumentations/src/mysql.js:84]:
```suggestion
    const sql = arguments[0].sql || arguments[0]
```

File: packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js [hunk 8]
```diff
@@ -8,52 +11,123 @@ const vulnerabilityReporter = require('../../../../src/appsec/iast/vulnerability
 
 describe('sql-injection-analyzer with mysql', () => {
   let mysql
-  let connection
   withVersions('mysql', 'mysql', version => {
     prepareTestServerForIast('mysql', (testThatRequestHasVulnerability, testThatRequestHasNoVulnerability) => {
+      const srcFilePath = path.join(__dirname, 'resources', 'sql-injection-methods.js')
```
Comment (simon-id) [inline packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:16]: I'm not sure why we need to use temp folders and fs calls here. It's good practice to really avoid those as they add complexity when it's not always needed. Is there an alternative solution here ?

**Review by simon-id — COMMENTED** [review]
We should get approval of the tracer team too

File: packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js (hunk 8 above)
Comment (uurien) [reply packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:16]: We are doing this kind of things to check that we are detecting the vulnerability in the correct file/line.
As we are excluding all the files in `dd-trace` directory, the code in the repository is in excluded files, so we are moving some files to temporal directories to check that the file/line is correct.

Comment (simon-id) [reply packages/dd-trace/test/appsec/iast/analyzers/sql-injection-analyzer.mysql.plugin.spec.js:16]: Aaah, make sense now my bad :+1:

**Review by uurien — APPROVED** [review]

**Pull request merged**